File: Dsa-day6.cpp
class Solution {
public:
    string remaining;

    int removePattern(const string& s, char first, char second, int value) {
        string temp;
        int score = 0;

        for (char c : s) {
            if (!temp.empty() && temp.back() == first && c == second) {
                temp.pop_back();  
                score += value;   
            } else {
                temp.push_back(c);
            }
        }

        remaining = temp;
        return score;
    }

    int maximumGain(string s, int x, int y) {
        int totalScore = 0;

        if (x > y) {
            totalScore += removePattern(s, 'a', 'b', x);
            totalScore += removePattern(remaining, 'b', 'a', y);
        } else {
            totalScore += removePattern(s, 'b', 'a', y);
            totalScore += removePattern(remaining, 'a', 'b', x);
        }

        return totalScore;
    }
};

File: Dsa-day6.py
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        def remove_pattern(s: str, first: str, second: str, score: int) -> (str, int):
            stack = []
            total = 0
            for c in s:
                if stack and stack[-1] == first and c == second:
                    stack.pop()
                    total += score
                else:
                    stack.append(c)
            return ''.join(stack), total

        total_score = 0
        if x > y:
            s, score1 = remove_pattern(s, 'a', 'b', x)
            _, score2 = remove_pattern(s, 'b', 'a', y)
        else:
            s, score1 = remove_pattern(s, 'b', 'a', y)
            _, score2 = remove_pattern(s, 'a', 'b', x)
        
        return score1 + score2
